src/app/CASESessionManager.cpp

```diff
@@ -0,0 +1,208 @@
+    OperationalDeviceProxy * session = FindSession(deviceId);
+    if (session == nullptr)
+    {
+        session = mActiveSessions.CreateObject(mInitParams.sessionInitParams, peerId);
+        if (session == nullptr)
+        {
+            onFailure->mCall(onFailure->mContext, deviceId, CHIP_ERROR_NO_MEMORY);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+    }
+
+    bool validAddress = (address != Transport::PeerAddress::UDP(Inet::IPAddress::Any));
+
+    if (!validAddress && mInitParams.dnsCache != nullptr)
+    {
+        Dnssd::ResolvedNodeData nodeData;
+        if (mInitParams.dnsCache->Lookup(peerId, nodeData) == CHIP_NO_ERROR)
+        {
+            address      = ToPeerAddress(nodeData);
+            validAddress = true;
+        }
+    }
+
+    if (validAddress)
+    {
+        uint32_t idleInterval;
+        uint32_t activeInterval;
+        session->GetMRPIntervals(idleInterval, activeInterval);
+        CHIP_ERROR err = session->UpdateDeviceData(address, idleInterval, activeInterval);
+        if (err != CHIP_NO_ERROR)
+        {
+            ReleaseSession(session);
+            return err;
+        }
+    }
+
+    CHIP_ERROR err = session->Connect(onConnection, onFailure);
+    if (err != CHIP_NO_ERROR)
+    {
+        ReleaseSession(session);
+    }
+
+    return err;
+}
+
+void CASESessionManager::ReleaseSession(NodeId deviceId)
+{
+    ReleaseSession(FindSession(deviceId));
+}
+
+CHIP_ERROR CASESessionManager::ResolveDeviceAddress(NodeId deviceId)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+    return Dnssd::Resolver::Instance().ResolveNodeId(mInitParams.sessionInitParams.fabricInfo->GetPeerIdForNode(deviceId),
+                                                     Inet::IPAddressType::kAny);
+}
+
+Transport::PeerAddress CASESessionManager::ToPeerAddress(const Dnssd::ResolvedNodeData & nodeData) const
+{
+    Inet::InterfaceId interfaceId;
+
+    // Only use the mDNS resolution's InterfaceID for addresses that are IPv6 LLA.
+    // For all other addresses, we should rely on the device's routing table to route messages sent.
+    // Forcing messages down an InterfaceId might fail. For example, in bridged networks like Thread,
+    // mDNS advertisements are not usually received on the same interface the peer is reachable on.
+    if (nodeData.mAddress[0].IsIPv6LinkLocal())
+    {
+        interfaceId = nodeData.mInterfaceId;
+    }
+
+    return Transport::PeerAddress::UDP(nodeData.mAddress[0], nodeData.mPort, interfaceId);
+}
+
+void CASESessionManager::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    VerifyOrReturn(mInitialized);
+
+    if (mInitParams.dnsCache != nullptr)
+    {
+        LogErrorOnFailure(mInitParams.dnsCache->Insert(nodeData));
+    }
+
+    OperationalDeviceProxy * session = FindSession(nodeData.mPeerId.GetNodeId());
+    VerifyOrReturn(session != nullptr);
```

Comment:
is it expected to not find such sessions? Should we log? silent return sounds as if we would have a hard time debugging issues. Same question for the mInitialized check above: at what time do we resolve node ids before being initialized?

Reply:
I'll add the log.

In the current code base, the node ID resolution is triggered from multiple places in the code. But, there's only one delegate, that receives all the callbacks. So in some use-case, if a session is not up with a node, but some controller app (for example) triggered the resolution of IP of a node, this function will get called.